zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java

```diff
@@ -2800,26 +2807,100 @@ public void setACL(final String path, List<ACL> acl, int version, StatCallback c
+            response = new GetChildrenPaginatedResponse();
+            ReplyHeader r = cnxn.submitRequest(h, request, response, wcb);
+            if (r.getErr() != 0) {
+                throw KeeperException.create(KeeperException.Code.get(r.getErr()),
+                        path);
+            }
+            minCzxId = response.getNextPageCzxid();
+            czxidOffset = response.getNextPageCzxidOffset();
+            needNextPage = needNextPage(maxReturned, minCzxId, czxidOffset);
+
+            if (isFirstPage) {
+                // If all children are returned in the first page,
+                // no need to use hash set to de-dup children
+                if (!needNextPage) {
+                    updateNextPage(nextPage, minCzxId, czxidOffset);
+                    return response.getChildren();
+                }
+                children = new HashSet<>();
+                isFirstPage = false;
+            }
+
+            children.addAll(response.getChildren());
+        }
+
+        updateNextPage(nextPage, minCzxId, czxidOffset);
+
+        return new ArrayList<>(children);
+    }
+
+    /**
+     * Returns a list of all the children given the path.
+     * <p>
+     * The difference between this API and {@link #getChildren(String, boolean)} is:
+     * when there are lots of children and the network buffer exceeds {@code jute.maxbuffer},
+     * this API will fetch the children using pagination and be able to return all children;
+     * while {@link #getChildren(String, boolean)} will fail.
+     * <p>
+     * The final list of children returned is NOT strongly consistent with the server's data:
+     * the list might contain some deleted children if some children are deleted before the last page is fetched.
+     * <p>
+     * If the watch is true and the call is successful (no exception is thrown),
+     # a watch will be left on the node with the given path. The watch will be
+     * triggered by a successful operation that deletes the node of the given
+     * path or creates/deletes a child under the node.
+     * <p>
+     *
+     * @param path the path of the parent node
+     * @param watch whether or not leave a watch on the given node
+     * @return a list of all children of the given path
+     * @throws KeeperException if the server signals an error with a non-zero error code.
+     * @throws InterruptedException if the server transaction is interrupted.
+     */
+    public List<String> getAllChildrenPaginated(String path, boolean watch)
+            throws KeeperException, InterruptedException {
+        return getChildren(path, watch ? watchManager.defaultWatcher : null, Integer.MAX_VALUE, 0L, 0, null);
+    }
+
+    private boolean needNextPage(int maxReturned, long minCzxId, int czxIdOffset) {
+        return maxReturned == Integer.MAX_VALUE
+                && minCzxId != ZooDefs.GetChildrenPaginated.lastPageMinCzxid
+                && czxIdOffset != ZooDefs.GetChildrenPaginated.lastPageMinCzxidOffset;
```

Comment:
If maxReturned is not MAX, then we always not getting the next page? It seems to be the wrong logic.

Reply:
When minCzxId != lastPageMinCzxid but czxIdOffset == lastPageMinCzxidOffset, the logical meaning of the condition might be confusing. Some may think the next page is still needed, since the minCzxId. But the result is false.
I suggest that we respect minCzxId only. If czxIdOffset is not a valid one, just treat it as zero. Will it conflict with some of the serverside assumptions?

Reply:
1. What I consider is, if users directly call this paginated API, it is supposed to be only one page (same as the original patch). Either the batchSize children would be returned, or less children than batchSize, and the client should use the NextPage info to check if it is the last page or fetch next page.
I'd like to make it transparent and only one request is submitted as users specify: no multiple requests/pages are made. If multi pages are returned for a batchSize < MAX, I feel that it is not transparent to users. Well, the MAX case is a special case to fetch all children.

2. If we want to always return the batchSize children, even with more than 1 page, we could do that. We just need to agree on the behavior.

3. I felt that setting both to -1 is stricter. But it's OK to just set `minCzxid` and treat it as the last page.

Reply:
"setting both to -1 is stricter." I agree. But how do you enforce our users to always set both -1 in their code? My point is that our API needs to tolerate different inputs in a graceful way.

And I didn't get the point 1 and 2.
What I meant is that,
1. maxReturned == MAX, then we are fetching multiple pages. Fine.
2. maxReturned == 10K, for example, then we are only fetching the first page (which may contain 400 nodes only). If the server side has a folder with children count 800, then the result is very strange. We shall still get the next page and fill the result with all 800 nodes, right?

Reply:
For 2, maxReturned = 10K, if only 400 nodes are returned in the 1st page, the `nextPage` will contain the minCzxid and offset of the next page for the users to fetch the remaining children, as you said, to fill the result with all 800 nodes.

Native ZK doesn't automatically do multiple pages, but the client can have the logic, similar as  `getAllChildrenPaginated()` in Zookeeper.java